mm/pmm.c

```diff
@@ -674,3 +748,41 @@ void map_frames_array(void) {
         BUG_ON(!vmap_kern_4k(va, mfn, L1_PROT));
     }
 }
+
+/* functions for paging to avoid deadlocks */
+
+frame_t *get_free_frame_norefill(void) {
+    frame_t *frame;
+
+    spin_lock(&priority_lock);
+    frame = reserve_frame(get_first_frame(free_frames, PAGE_ORDER_4K));
+    spin_unlock(&priority_lock);
+
+    /* we ran out of reserved frames. increase MIN_NUM_4K_FRAMES */
+    BUG_ON(!frame);
+
+    return frame;
+}
+
+void refill_from_paging(void) {
+    spin_lock(&priority_lock);
+
+    /* avoid refill_from_paging being called from refill_from_paging */
+    if (!refill_from_paging_ongoing) {
```

Comment:
since this is the only scope where this variable is being used, declare it here as a `static` variable.

Reply:
Also, I think this variable should be declared as an atomic one and handled via atomic functions

Reply:
I just realized that since the caller of this function has to hold the vmap_lock we can move this check outside of the spin lock and don't need to use atomics either.